subsys/bluetooth/audio/stream.c

```diff
@@ -0,0 +1,1100 @@
+		    c->ep->cis_id == info->cis_id) {
+			*iso_chan = enabling[i]->iso;
+			enabling[i] = NULL;
+			return 0;
+		}
+	}
+
+	BT_ERR("No channel listening");
+
+	return -EPERM;
+}
+
+static struct bt_iso_server iso_server = {
+	.sec_level = BT_SECURITY_L2,
+	.accept = bt_audio_stream_iso_accept,
+};
+
+int bt_audio_stream_iso_listen(struct bt_audio_stream *stream)
+{
+	static bool server;
+	int err, i;
+	struct bt_audio_stream **free_stream = NULL;
+
+	BT_DBG("stream %p conn %p", stream, stream->conn);
+
+	if (server) {
+		goto done;
+	}
+
+	err = bt_iso_server_register(&iso_server);
+	if (err) {
+		BT_ERR("bt_iso_server_register: %d", err);
+		return err;
+	}
+
+	server = true;
+
+done:
+	for (i = 0; i < ARRAY_SIZE(enabling); i++) {
+		if (enabling[i] == stream) {
+			return 0;
+		}
+
+		if (enabling[i] == NULL && free_stream == NULL) {
+			free_stream = &enabling[i];
+		}
+	}
+
+	if (free_stream != NULL) {
+		*free_stream = stream;
+		return 0;
+	}
+
+	BT_ERR("Unable to listen: no slot left");
+
+	return -ENOSPC;
+}
+
+int bt_audio_codec_qos_to_iso_qos(struct bt_iso_chan_qos *qos,
+				  struct bt_codec_qos *codec)
+{
+	struct bt_iso_chan_io_qos *io;
+
+	switch (codec->dir) {
+	case BT_CODEC_QOS_IN:
+		io = qos->rx;
+		break;
+	case BT_CODEC_QOS_OUT:
+		io = qos->tx;
+		break;
+	case BT_CODEC_QOS_INOUT:
+		io = qos->rx = qos->tx;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	io->sdu = codec->sdu;
+	io->phy = codec->phy;
+	io->rtn = codec->rtn;
+
+	return 0;
+}
+
+bool bt_audio_valid_qos(const struct bt_codec_qos *qos)
+{
+	if (qos->interval < BT_ISO_INTERVAL_MIN ||
+	    qos->interval > BT_ISO_INTERVAL_MAX) {
+		BT_DBG("Interval not within allowed range: %u (%u-%u)",
+		       qos->interval, BT_ISO_INTERVAL_MIN, BT_ISO_INTERVAL_MAX);
+		return false;
+	}
+
+	if (qos->framing > BT_CODEC_QOS_FRAMED) {
+		BT_DBG("Invalid Framing 0x%02x", qos->framing);
+		return false;
+	}
+
+	if (qos->phy != BT_CODEC_QOS_1M &&
+	    qos->phy != BT_CODEC_QOS_2M &&
+	    qos->phy != BT_CODEC_QOS_CODED) {
+		BT_DBG("Invalid PHY 0x%02x", qos->phy);
+		return false;
+	}
+
+	if (qos->sdu > BT_ISO_MAX_SDU) {
+		BT_DBG("Invalid SDU %u", qos->sdu);
+		return false;
+	}
+
+	if (qos->latency < BT_ISO_LATENCY_MIN ||
+	    qos->latency > BT_ISO_LATENCY_MAX) {
+		BT_DBG("Invalid Latency %u", qos->latency);
+		return false;
+	}
+
+	return true;
+}
+
+static bool bt_audio_stream_is_broadcast(const struct bt_audio_stream *stream)
+{
+	/* TODO: Add broadcast sink */
+	return (IS_ENABLED(CONFIG_BT_AUDIO_BROADCAST_SOURCE) &&
+		bt_audio_ep_is_broadcast_src(stream->ep)) ||
+	       (IS_ENABLED(CONFIG_BT_AUDIO_BROADCAST_SINK) &&
+		bt_audio_ep_is_broadcast_snk(stream->ep));
+}
+
+bool bt_audio_valid_stream_qos(const struct bt_audio_stream *stream,
+			       const struct bt_codec_qos *qos)
+{
+	const struct bt_codec_qos_pref *qos_pref = &stream->ep->qos_pref;
+
+	if (qos_pref->latency < qos->latency) {
+		BT_DBG("Latency %u higher than preferred max %u",
```

Comment:
I agree that this could be done differently. It's not a choice I've made (existing code), but I do see the point of not letting the upper layers set values that are outside the preference. We could let this be up to the upper layers completely though.

Reply:
We'll leave it for now.